docs/api_specs/includes/watcher_api_specs.md

```diff
@@ -0,0 +1,650 @@
+</aside>
+
+### Byzantine events
+All of the following events indicate byzantine behaviour and that the user should either exit or challenge.
+
+#### `invalid_exit`
+> An invalid_exit event
+
+```json
+{
+    "event": "invalid_exit",
+    "details": {
+        "eth_height"  : 3521678,
+        "utxo_pos"  : 10001001,
+        "owner"  : "B3256026863EB6AE5B06FA396AB09069784EA8EA",
+        "currency"  : "0000000000000000000000000000000000000000",
+        "amount" : 100
+    }
+}
+```
+
+Indicates that an invalid exit is occurring. It should be challenged.
+
+
+#### `unchallenged_exit`
+> An unchallenged_exit event
+
+```json
+{
+    "event": "unchallenged_exit",
+    "details": {
+        "eth_height"  : 3521678,
+        "utxo_pos"  : 10001001,
+        "owner"  : "B3256026863EB6AE5B06FA396AB09069784EA8EA",
+        "currency"  : "0000000000000000000000000000000000000000",
+        "amount" : 100
+    }
+}
+```
+
+Indicates that an invalid exit is dangerously close to finalization and hasn't been challenged. User should exit.
+
+
+#### `invalid_block`
+> An invalid_block event
+
+```json
+{
+    "event": "invalid_block",
+    "details": {
+        "blockhash"  : "DB32876CC6F26E96B9291682F3AF4A04C2AA2269747839F14F1A8C529CF90225",
+        "blocknum"  : 10000,
+    }
+}
+```
+
+An invalid block has been added to the chain. User should exit.
+
+
+#### `block_withholding`
+> A block_withholding event
+
+```json
+{
+    "event": "block_withholding",
+    "details": {
+        "blockhash"  : "DB32876CC6F26E96B9291682F3AF4A04C2AA2269747839F14F1A8C529CF90225",
+        "blocknum"  : 10000,
+    }
+}
+```
+
+The ChildChain is withholding a block whose hash has been published on the root chain. User should exit.
+
+#### `noncanonical_ife`
+> A noncanonical_ife event
+
+```json
+{
+    "event": "noncanonical_ife",
+    "details": {
+        "txbytes": "F3170101C0940000..."
+    }
+}
+```
+
+An in-flight exit of a non-canonical transaction has been started. It should be challenged. 
+<aside class="warning"> Not Implemented Yet.</aside> 
+
+Event details:
+Attribute | Type | Description
+--------- | ------- | -----------
+txbytes | Hex encoded string | The in-flight transaction that the event relates to
+
+#### `invalid_ife_challenge`
+> A invalid_ife_challenge event
+
+```json
+{
+    "event": "invalid_ife_challenge",
+    "details": {
+        "txbytes": "F3170101C0940000..."
+    }
+}
+```
+
+A canonical in-flight exit has been challenged. The challenge should be responded to. 
+<aside class="warning"> Not Implemented Yet.</aside> 
+
+Event details:
+Attribute | Type | Description
+--------- | ------- | -----------
+txbytes | Hex encoded string | The in-flight transaction that the event relates to
+
+#### `piggyback_available`
+> A piggyback_available event
+
+```json
+{
+    "event": "piggyback_available",
+    "details": {
+        "txbytes": "F3170101C0940000...",
+        "available_outputs" : [0, 1],
+        "available_inputs" : [0, 1]
+    }
```

Comment:
I think possible values for:
`available_outputs` 4, 5, 6,7 
`available_inputs`: 0, 1, 2, 3
Maybe it is worth to  mention somewhere here?

Reply:
note that the `outputs` being numbered "on top" of `inputs` (4,5,6,7) should probably be seen as leakage of contract implementation details.

Reply:
Yeah, I would prefer to keep them 0-based in the API, and not expose the contract implementation details.

Reply:
So the flow will be the following:
1.  A User gets `piggyback_available` event with "available_outputs" : [0, 1] 
2.  A User calls `get_piggyback_data`  
3.  A user wants to make piggyback for output 1
4.  A user must manually add 4 to output 1.
5.  A user calls root chain.`piggybackInFlightExit` method with `_inFlightTx` + `_outputIndex` = 5

Am I missing something ? 

note:
```
 function piggybackInFlightExit(
        bytes _inFlightTx,
        uint8 _outputIndex
    )
```

Reply:
Correct, except for `2.` isn't necessary anymore because all the info necessary to piggyback is already in the event.

I know that manual +4 is ugly, but that's how the contract works. We could try to hide that from the user by automatically doing the +4 for them, but I feel it makes the API more brittle.
0-based arrays feel more neutral somehow.

I'm open to being persuaded otherwise though...

Reply:
I would pitch doing `0-based and abstractly correct` in `status/events` and `contract-friendly` in the return of "get-piggyback-information" endpoint (to adhere by the conventions), but isn't this the endpoint that we just dubbed redundant and removed :joy: ? (Sorry if I'm lost, I'm not fully following the thread, feel free to disregard comment).

Reply:
Yes, I removed 'get-piggyback-information' because the only info you need are `txbytes` and the (controversial) index.